rustls/src/conn.rs

```diff
@@ -649,31 +651,40 @@ impl<Data> ConnectionCore<Data> {
             }
         };
 
-        let mut borrowed_buffer = deframer_buffer.borrow();
-        while let Some(msg) = self.deframe(Some(&*state), &mut borrowed_buffer)? {
+        let mut discard = 0;
+        loop {
+            let mut borrowed_buffer = deframer_buffer.borrow();
+            borrowed_buffer.queue_discard(discard);
+
+            let opt_msg = self.deframe(Some(&*state), &mut borrowed_buffer)?;
```

Comment:
I think this continues to have the same issue as the other PR? `borrowed_buffer.pending_discard()` is crucial data that must not be lost on error.

Reply:
maybe here we can just match on the output of deframe and discard on error. What do you think?

Reply:
I think it would be more robush if `deframer_buffer.borrow()` returned an object that automatically took care of this on drop. However, I think achieving that is not currently possible because the deframer buffer is mutable (despite deframing being a strictly read-only operation) and fixing that is a larger refactor.

So I guess ensuring every caller of `deframer_buffer.borrow()` does the right thing is the second best option now.

Reply:
Ok I did the second best thing :P